Sprint 8 solution time and duration

shinket7 opened the pull request:
Добавил работу со временем по задачам. Добавил новые тесты.

shinket7 committed:
Add start time and duration to `Task`.

Add fields: `startTime`, `duration`.
Add methods: `getStartTime()`, `getDuration()`, `setStartTimeAndDuration()`, `getEndTime()`.

shinket7 committed:
Add epic start time and duration computing.

shinket7 committed:
Add time start and duration fields to autosave file.

shinket7 committed:
Add time start and duration fields to autosave file.

Change tests to new features.

shinket7 committed:
Add prioritized tasks field. Add `compareTo()` to `Task`

shinket7 committed:
Add deletions from time periods of 15 minutes.

shinket7 committed:
Add additions to time periods of 15 minutes.

shinket7 committed:
Create task overlapping check.

shinket7 committed:
Add task overlapping check to tasks add and update.

shinket7 committed:
Add tests for `Task` methods.

shinket7 committed:
Create abstract 'TaskManagerTest' class.

Move all same tests of managers to one abstract class.

shinket7 committed:
Fix demo file.

shinket7 committed:
Add tests of overlapped tasks.

avfyodorov commented on src/tracker/controllers/InMemoryTaskManager.java:
```diff
@@ -254,14 +400,52 @@ private TaskStatus computeEpicStatus(int epicId) {
+        if (firstTimeStart == null) {
+            return;
+        }
+        epic.setStartTimeAndDuration(firstTimeStart, Duration.between(firstTimeStart, lastTimeEnd));
```
Duration.between(firstTimeStart, lastTimeEnd))

По техническому заданию - "Продолжительность эпика — сумма продолжительностей всех его подзадач. "
А так считается разница между первой и последней.

shinket7 replied:
Изменил логику расчёта продолжительности эпика на сумму продолжительностей его задач.

avfyodorov commented on src/tracker/controllers/InMemoryTaskManager.java:
```diff
@@ -254,14 +400,52 @@ private TaskStatus computeEpicStatus(int epicId) {
+            if (lastTimeStart == null || lastTimeStart.isBefore(subtaskTimeStart)) {
+                lastTimeStart = subtaskTimeStart;
+                final Duration duration = Duration.between(firstTimeStart, lastTimeStart.plus(subtask.getDuration()));
+                lastTimeEnd = lastTimeStart.plus(duration);
```
lastTimeEnd = lastTimeStart.plus(duration);

Ещё нужно рассчитать и установить дату окончания Эпика. Обратите внимание, что дата окончания определяется по-другому, не так как в обычной задаче.
.---
время завершения — время окончания самой поздней из задач.

shinket7 replied:
Сделал тут запись даты окончания эпика в экземпляр эпика.

avfyodorov commented on src/tracker/controllers/InMemoryTaskManager.java:
```diff
@@ -254,14 +400,52 @@ private TaskStatus computeEpicStatus(int epicId) {
+        final Epic epic = epics.get(epicId);
+        final ArrayList<Integer> subtaskIds = epic.getSubtaskIds();
+
+        if (subtaskIds.isEmpty()) {
```
Тут можно было бы обнулить параметры Epica перед выходом из метода. 
Такой вариант - были у Эпика подзадачи (Время и продолжительность рассчитаны), потом их удалили (нужно обновить параметры Epica), а параметры Эпика остались.

shinket7 replied:
Добавил обнуление полей начала, продолжительности, окончания эпика.

avfyodorov commented on src/tracker/model/Task.java:
```diff
@@ -1,19 +1,25 @@
 import java.util.Objects;
 
-public class Task {
+public class Task implements Comparable<Task> {
```
Пишу здесь, Потому что класс Epic недоступен для комментария. 

По ТЗ:
.-------
Для реализации getEndTime() удобно добавить поле endTime в Epic и рассчитать его вместе с другими полями.

shinket7 replied:
Сделал в эпике поле со временем завершения. Сделал метод для записи этого поля, а также переопределил метод `getEndTime()`.

avfyodorov commented on src/tracker/controllers/InMemoryTaskManager.java:
```diff
@@ -2,23 +2,27 @@
     private final HashMap<Integer, Epic> epics;
     private final HashMap<Integer, Subtask> subtasks;
+    private final Set<Task> prioritizedTasks;
+    private final TreeMap<LocalDateTime, Collection<Task>> occupiedPeriod15StampsAndTasks;
```
`    private final TreeMap<LocalDateTime, Collection<Task>> occupiedPeriod15StampsAndTasks;`

Конечно, на Ваше усмотрение, но... 
По-моему, это лишняя структура, есть prioritizedTasks, которая содержит все задачи и подзадачи. 
Можно добавить компаратор и работа упрощается 

`	protected final Set<Task> prioritizedTasks = new TreeSet<>(Comparator.comparing(Task::getStartTime));`

shinket7 replied:
Насколько понимаю, тут не получится заменить структуру с `occupiedPeriod15StampsAndTasks` через `prioritizedTasks` — они отвечают за разный функционал.

- `prioritizedTasks` отвечает за сортировку задач по времени их старта. Сортировка происходит через `Comparable` интерфейс у класса `Task`.
- `occupiedPeriod15StampsAndTasks` отвечает за то, что таски и сабтаски не пересекаются по времени друг с другом. Так как задачи могут быть начаты в любой момент времени и иметь любую продолжительность времени, то не получится учитывать только время начала, даже при приведении к 15-ти минутным промежуткам.

Идея логики с `occupiedPeriod15StampsAndTasks` в том, чтобы иметь оптимизированную по времени выполнения точную проверку пересечения задач с произвольным временем старта и продолжительностью. В моём понимании деление по 15-ти минутным промежуткам занимает O(1) и сокращает количество задач для проверки до нескольких. Но при этом если в одном таком промежутке заканчивается одна задача и начинается другая, то будет точная проверка этих задач, что они не пересекаются, за O(n). Но так как таких задач останется мало, то проверка должна быть всё равно близкой к O(1).

avfyodorov replied:
Тут, видите, какая штука. 
В техническом задании нет требования выполнять за О(1). Только за O(n).
.---
 Предполагается, что пользователь будет часто запрашивать этот список, поэтому подберите подходящую структуру данных для хранения. Сложность получения должна быть уменьшена …. до O(n)
.---

То есть получается, что достаточно иметь одну структуру (TreeSet),  в ней хранить задачи, отсортированные по времени и здесь же проверять на пересечение (в цикле при добавлении / обновлении).
Вот, например, вполне рабочий вариант. Одна структура, один цикл, время O(n) 
```java
//Приватный метод менеджера задач. Вызывается при добавлении и обновлении задач и подзадач.
	private void add(Task task) {
		final LocalDateTime startTime = task.getStartTime();
		final LocalDateTime endTime = task.getEndTime();
		for (Task t : prioritizedTasks) {
			final LocalDateTime existStart = t.getStartTime();
			final LocalDateTime existEnd = t.getEndTime();
//Если задача, которую нужно добавить, заканчивается раньше, чем начинается очередная Задача в списке,
// то всё в порядке, они не пересекаются.
			if (!endTime.isAfter(existStart)) {// newTimeEnd <= existTimeStart
				continue;
			}
//Если очередная задача из списка заканчивается раньше, чем начинается добавляемая задача,
// тоже всё в порядке, пересечений нет.
			if (!existEnd.isAfter(startTime)) {// existTimeEnd <= newTimeStart
				continue;
			}

			throw new TaskValidationException("Задача пересекаются с id=" + t.getId() + " c " + existStart + " по " + existEnd);
		}

		prioritizedTasks.add(task);
	}
```

shinket7 replied:
По итогу решил убрать эту оптимизацию поиска пересечений. По ТЗ она необязательна, но при этом она существенно усложняет понимание кода программы из-за своего объёма.

avfyodorov reviewed (commented):
Добрый день, Виталий! 

Хорошая работа, всего несколько небольших уточнений.

avfyodorov commented on src/tracker/controllers/InMemoryTaskManager.java:
```diff
@@ -52,10 +56,47 @@ public ArrayList<Integer> getSubtaskIds() {
     }
 
+    private LocalDateTime getPeriod15Stamp(LocalDateTime dateTime) {
+        final int remainder = dateTime.getMinute() % 15;
```
15 ????????

Всё-таки лучше обойтись без констант жёстко зашитых внутрь кода.

shinket7 replied:
В смысле при данной реализацией с усложнённым алгоритмом поиска пересечений стоило бы вынести это значение для деления в переменную-константу с говорящим названием и её указать где-то в начале? Или откуда это значение лучше тянуть?

shinket7 committed:
Change `Epic` duration calculation.

Now `Epic` duration equals to sum of durations of its subtasks.
Add field 'endTime' to `Epic` to remove repeated calculations.

shinket7 committed:
Add test of epic duration calculation.

Test checks that epic's duration equals to sum of its subtask durations.

shinket7 committed:
Add epic time fields reset in computing if no subtasks.

shinket7 committed:
Add test of getting prioritized tasks.

avfyodorov reviewed (commented):
Добрый день, Виталий!

Замечаний нет.
Работа принята.
Оставил комментарий на гитхабе, посмотрите, пожалуйста.

shinket7 committed:
Specify `shouldReturnSortedByStartTimeIssues()` test.

shinket7 committed:
Delete complicated logic of issue overlapping search.

Remove optimization which made the difficulty near O(1) as it's too complicated but not needed now.
